File: ui/src/models/task.ts
/**
 * Task Domain Model
 * Core entity for task management with value objects
 */

// Task status: dynamic, read from config.settings.statuses
export type TaskStatus = string;

// Priority: low, medium, high
export type TaskPriority = "low" | "medium" | "high";

// Task interface
export interface Task {
	id: string; // task-{6_char_base36} for new tasks; legacy sequential IDs still supported
	title: string;
	description?: string;
	status: TaskStatus;
	priority: TaskPriority;
	assignee?: string; // "@harry"
	labels: string[];
	parent?: string; // Parent task ID for subtasks
	subtasks: string[]; // Child task IDs
	spec?: string; // Linked spec document path (e.g., "specs/user-auth" for @doc/specs/user-auth)
	fulfills?: string[]; // Spec ACs this task fulfills (e.g., ["AC-1", "AC-2"])
	order?: number; // Manual ordering for display (lower = first)
	createdAt: Date;
	updatedAt: Date;

	// Acceptance criteria
	acceptanceCriteria: AcceptanceCriterion[];

	// Time tracking
	timeSpent: number; // Total seconds
	timeEntries: TimeEntry[];

	// Notes
	implementationPlan?: string;
	implementationNotes?: string;
}

export interface AcceptanceCriterion {
	text: string;
	completed: boolean;
}

export interface TimeEntry {
	id: string;
	startedAt: Date;
	endedAt?: Date;
	duration: number; // Seconds
	note?: string;
}

// Helper functions for task creation
export function createTask(
	data: Omit<Task, "id" | "createdAt" | "updatedAt" | "subtasks" | "timeSpent" | "timeEntries">,
): Omit<Task, "id"> {
	const now = new Date();
	return {
		...data,
		subtasks: [],
		timeSpent: 0,
		timeEntries: [],
		createdAt: now,
		updatedAt: now,
	};
}

// Default statuses (fallback if config not available)
export const DEFAULT_STATUSES: TaskStatus[] = [
	"todo",
	"in-progress",
	"in-review",
	"done",
	"blocked",
	"on-hold",
	"urgent",
];

// Helper to validate task status
// Use with allowed statuses from config, or falls back to DEFAULT_STATUSES
export function isValidTaskStatus(status: string, allowedStatuses?: TaskStatus[]): status is TaskStatus {
	const validStatuses = allowedStatuses || DEFAULT_STATUSES;
	return validStatuses.includes(status);
}

// Helper to validate task priority
export function isValidTaskPriority(priority: string): priority is TaskPriority {
	return ["low", "medium", "high"].includes(priority);
}

File: internal/models/task.go
package models

import "time"

// Task represents a task in the .knowns/ system.
// The ID uses a 6-character base36 format (e.g., "abc123").
// Legacy sequential integer IDs are also supported for backward compatibility.
type Task struct {
	ID          string   `json:"id"                    yaml:"id"`
	Title       string   `json:"title"                 yaml:"title"`
	Description string   `json:"description,omitempty" yaml:"description,omitempty"`
	Status      string   `json:"status"                yaml:"status"`
	Priority    string   `json:"priority"              yaml:"priority"` // "low", "medium", "high"
	Assignee    string   `json:"assignee,omitempty"    yaml:"assignee,omitempty"`
	Labels      []string `json:"labels"                yaml:"labels"`

	// Parent task ID for subtasks (e.g., "abc123").
	Parent string `json:"parent,omitempty" yaml:"parent,omitempty"`

	// Subtasks holds child task IDs. Derived at load time; not persisted in the
	// task file itself.
	Subtasks []string `json:"subtasks,omitempty" yaml:"-"`

	// Spec is the linked spec document path (e.g., "specs/user-auth").
	Spec string `json:"spec,omitempty" yaml:"spec,omitempty"`

	// Fulfills lists spec acceptance-criteria IDs this task satisfies
	// (e.g., ["AC-1", "AC-2"]).
	Fulfills []string `json:"fulfills,omitempty" yaml:"fulfills,omitempty"`

	// Order is an optional manual display order (lower = first).
	Order *int `json:"order,omitempty" yaml:"order,omitempty"`

	CreatedAt time.Time `json:"createdAt" yaml:"createdAt"`
	UpdatedAt time.Time `json:"updatedAt" yaml:"updatedAt"`

	// AcceptanceCriteria lives in the markdown body, not the YAML frontmatter.
	AcceptanceCriteria []AcceptanceCriterion `json:"acceptanceCriteria" yaml:"-"`

	// TimeSpent is the total time tracked in seconds.
	TimeSpent   int         `json:"timeSpent"            yaml:"timeSpent"`
	TimeEntries []TimeEntry `json:"timeEntries,omitempty" yaml:"-"`

	// ImplementationPlan and ImplementationNotes live in the markdown body.
	ImplementationPlan  string `json:"implementationPlan,omitempty"  yaml:"-"`
	ImplementationNotes string `json:"implementationNotes,omitempty" yaml:"-"`
}

// AcceptanceCriterion is a single checkbox item in the Acceptance Criteria
// section of a task file.
type AcceptanceCriterion struct {
	Text      string `json:"text"`
	Completed bool   `json:"completed"`
}

// DefaultStatuses returns the built-in task status list used when no project
// config is available.
func DefaultStatuses() []string {
	return []string{
		"todo",
		"in-progress",
		"in-review",
		"done",
		"blocked",
		"on-hold",
		"urgent",
	}
}

// ValidPriorities returns the allowed priority values.
func ValidPriorities() []string {
	return []string{"low", "medium", "high"}
}
